hands-on-exercise/4-run-in-prod/3-migration-prod.md

```diff
@@ -0,0 +1,618 @@
+
+## Prepare the Cosmovisor executable
+
+Because the project in its current state uses Cosmos SDK v0.45.4, to avoid any surprise you will prepare Cosmovisor at the [v0.45.4](https://docs.cosmos.network/v0.45/run-node/cosmovisor.html) version too.
+
+You can describe the steps in a new Dockerfile `prod-sim/Dockerfile-cosmovisor-alpine`, described here logically (before a recap lower down):
+
+1. You need to build Cosmovisor from its code:
+
+    ```Dockerfile [https://github.com/cosmos/b9-checkers-academy-draft/blob/migration-prod/prod-sim/Dockerfile-cosmovisor-alpine#L1-L25]
+    FROM --platform=linux golang:1.18.7-alpine AS builder
+
+    ENV COSMOS_VERSION=v0.45.4
+
+    RUN apk update
+    RUN apk add make git
+
+    WORKDIR /root
+    RUN git clone --depth 1 --branch ${COSMOS_VERSION} https://github.com/cosmos/cosmos-sdk.git
+
+    WORKDIR /root/cosmos-sdk/cosmovisor
+
+    RUN make cosmovisor
+
+    FROM --platform=linux alpine
+
+    ENV LOCAL=/usr/local
+
+    COPY --from=builder /root/cosmos-sdk/cosmovisor/cosmovisor ${LOCAL}/bin/cosmovisor
+    ```
+
+2. Cosmovisor is instructed via [environment variables](https://docs.cosmos.network/v0.45/run-node/cosmovisor.html#command-line-arguments-and-environment-variables). In the eventual containers, the `/root/.checkers` folder comes from a volume mount, so to avoid any conflict it is better not to put the `cosmovisor` folder directly inside it. Instead pick `/root/.checkers-upgrade`:
+
+    ```diff [https://github.com/cosmos/b9-checkers-academy-draft/blob/migration-prod/prod-sim/Dockerfile-cosmovisor-alpine#L20-L23]
+        ...
+        FROM --platform=linux alpine
+        
+        ENV LOCAL=/usr/local
+    +  ENV DAEMON_HOME=/root/.checkers-upgrade
+    +  ENV DAEMON_NAME=checkersd
+    +  ENV DAEMON_ALLOW_DOWNLOAD_BINARIES=false
+    +  ENV DAEMON_RESTART_AFTER_UPGRADE=true
+        ...
+    ```
+
+3. With the folder decided, you can introduce both the v1 and v2 checkers executables. They can be conveniently taken from their respective Docker images:
+
+    ```diff [https://github.com/cosmos/b9-checkers-academy-draft/blob/migration-prod/prod-sim/Dockerfile-cosmovisor-alpine#L15-L27]
+    +  FROM --platform=linux checkersd_i:v1-alpine AS v1
+    +  FROM --platform=linux checkersd_i:v2-alpine AS v2
+        FROM --platform=linux alpine
+        ...
+        COPY --from=builder /root/cosmos-sdk/cosmovisor/cosmovisor ${LOCAL}/bin/cosmovisor
+    +  COPY --from=v1 /usr/local/bin/checkersd $DAEMON_HOME/cosmovisor/genesis/bin/checkersd
+    +  COPY --from=v2 /usr/local/bin/checkersd $DAEMON_HOME/cosmovisor/upgrades/v1tov2/bin/checkersd
+        ...
+    ```
+
+    Checkers starts at v1, therefore the v1 code goes into `.../genesis`, and we know that the code of the eventual upgrade named `v1tov2` is the v2. Note also the decision to use `/usr/local` explicitly, as this is knowledge that is kept in a separate Docker image.
```

Comment:
```suggestion
Checkers starts at v1, therefore the v1 code goes into `.../genesis`, and we know that the code of the eventual upgrade named `v1tov2` is the v2. Note also the decision to use `/usr/local` explicitly, as this is knowledge that is kept in a separate Docker image.
```

Reply:
The space is to have this paragraph with the indentation of the numbered list, and also not break numbering.